### ISSUE TITLE
[FEA] Optionally replacing cub pool allocator with RMM at runtime.

### ISSUE THREAD

**Issue by trivialfis** [rapidsai/rmm#583]
**Is your feature request related to a problem? Please describe.**

Currently in [XGBoost](https://github.com/dmlc/xgboost) we build with optional support for RMM.  Internally XGBoost has 2 allocators when RMM is not used, first is from thrust for large trunk of data, which is just `cudaMalloc`.  Another one is from `cub` pool allocator for runtime memory.  We want to make sure that our pool allocator wrapper is always using memory pool instead of calling `cudaMalloc` for every allocation.  But if we use RMM as a replacement for cub, and RMM is not initialized by user, RMM will use `cudaMalloc` instead.

**Describe the solution you'd like**

There are a few options we can proceed.  More suggestions are welcomed!

1. A reliable way to detect whether memory pool is initialized or whether pool allocator in rmm is currently being used.  This way we can decide whether should the allocator fall back to cub at runtime.
2. A way for us to initialize a pool inside C++ code, that won't conflict with user setting.  That means, we will always initialize a pool in RMM, but if users have already done it, then RMM can intelligently use the pool requested by user and ignore our initialization.

**Comment by jrhemstad** [comment]
The problem with either of these options is that there is no one single "pool" implementation to detect. 

RMM provides several different high performance memory allocator implementations that can be used depending on the situation. What's more, a user may provide their own custom allocator implementation. On top of all that, there are various adaptors that can be layered on top of a resource, e.g., `logging_resource_adaptor`. 

Therefore, it's impossible to generically detect a situation where you should reinitialize RMM. Furthermore, I don't think a library should be trying to make decisions for a user like that. 

What we can and _should_ do is start defaulting cuDF to initialize RMM w/ `pool_memory_resource` as the default.

**Comment by trivialfis** [comment]
@jrhemstad Thanks for the reply.  What if `cuDF` is not used?  RMM is designed for other libraries and we might look at them individually.

**Comment by jrhemstad** [comment]
> @jrhemstad Thanks for the reply. What if `cuDF` is not used? RMM is designed for other libraries and we might look at them individually.

Yes, RMM is definitely designed to be used outside of cuDF. 

The problem is that what you're asking for isn't really possible. Essentially what you'd like is to answer the question "Should I replace the default resource?". But how do you answer that question? It's not enough to just check if the default is `pool_memory_resource`, as it could be wrapped in an adaptor, or another resource like `arena_memory_resource` could be used. Additionally, it's not enough to exhaustively check against all the resources RMM provides either as a user may have provided their own custom resource. 

One thing you _could_ do is check if the default resource is equal to `cuda_memory_resource`. If it is _not_ then that means the user overrode the default and set it to some custom resource. In that situation you probably don't want to mess with the default allocator. Even if `cuda_memory_resource` is the default, I don't think XGBoost should be automatically overriding it. A user may be purposefully using `cudaMalloc`, e.g., because they want to use `cuda-memcheck` to check for leaks. 

I think what would be most appropriate is to detect if `cuda_memory_resource` is the default and emit a warning. That way it is still the users choice and responsibility to set which resource is used while letting them know that `cuda_memory_resource` is going to make performance suffer.

**Comment by trivialfis** [comment]
> Even if cuda_memory_resource is the default, I don't think XGBoost should be automatically overriding it. 

That's a really good point.

> I think what would be most appropriate is to detect if cuda_memory_resource is the default and emit a warning.

That might not be feasible as pool allocator can be constructed a few hundred times a second.  Right now my preference is just defaulting to cub allocator for those allocations that must be cached, and use rmm in all the other places.

**Comment by jrhemstad** [comment]
> Right now my preference is just defaulting to cub allocator for those allocations that must be cached, and use rmm in all the other places, but others worry that might generate oom errors.

I don't think that is a good idea. If there are allocations that are always outside the view of RMM, that would defeat the purpose of integrating RMM into XGBoost. Not only would you be competing with RMM's pool and run the risk of OOM's, but RMM's implementations are faster than CUB and you'd be losing out on the performance too.

**Comment by trivialfis** [comment]
@jrhemstad  Em, I don't really have a solution right now.

- Detect whether RMM pool is initialized: Not possible to work reliably.
- Force initialization inside C++ code: a library should be trying to make decisions for a user like that.
- Default to RMM for everything: Forces user to understand and initialize memory pool.  Might generate too much warnings.  Also XGBoost will be competing with libraries that don't use RMM.
- Default to cub when allocating memory that must be cached: Might generate OOM.

**Comment by harrism** [comment]
Is this for the python interface or C++?

**Comment by trivialfis** [comment]
It's for using RMM in C++.

**Comment by hcho3** [comment]
@trivialfis @harrism @jrhemstad To be robust against all edge cases, we should ask the users to explicitly signal XGBoost that RMM is used. Something like:
```python
import xgboost

params = {'use_rmm': True, ... }

bst = xgboost.train(params, ...)
```

This will require substantial refactoring in XGBoost so that we can pass the hyperparameter `use_rmm` all the way to XGBoost's memory allocator logic.

**Comment by harrism** [comment]
@trivialfis said this was for C++ but @hcho3 gave a Python example (which is why I asked). :)

If XGBoost had an external allocator interface, then this would not be an issue. The user would just

```
import rmm

xgboost.set_allocator(rmm.xgboost_allocator)
```

This is exactly how RMM works with CUPy and Numba. See https://github.com/rapidsai/rmm/#using-rmm-with-cupy
and
https://github.com/rapidsai/rmm/#using-rmm-with-numba

This would require adding an external allocator interface to XGBoost, and some small additional Python APIs in RMM to configure and create the compatible allocator.  But the advantage is that it's up to the user to make it clear what allocator they want to use, and that choice is unambiguous. Moreover, it allows creating arbitrary custom allocators for XGBoost that have nothing to do with RMM (or even GPUs).

**Comment by hcho3** [comment]
@harrism Not sure if we can adopt that interface, given that we need to also support the JVM binding of XGBoost. That's why I proposed adding a new hyperparameter.

**Comment by harrism** [comment]
It's up to you, I don't have a horse in this race. I'm just suggesting a very generic and flexible approach that serves many libraries well. On the C++ side, STL and Thrust also allow providing a custom allocator in this way, except they do it at the container constructor and algorithm execution policy level, but that works well for C++ and those particular libraries. Same general idea.

**Comment by JohnZed** [comment]
@hcho3 - how heavy is the work to add a "use_rmm" option? Like many days of work and a lot of files touched, or is there a simper path?

Like everyone else, I don't _love_ any of these options, but the explicit parameter may be the least problematic solution. Not sure it is possible in time for the next release, but we should at least think it through.

**Comment by trivialfis** [comment]
@JohnZed Need to pass it all the way down to internal global allocator.  Also need to make sure the configuration works after serialization.

> how heavy is the work to add a "use_rmm" option? 

Moderate.

**Comment by harrism** [comment]
Should this issue be moved to the xgboost repo?

**Comment by hcho3** [comment]
@harrism Feel free to close this issue. I just created https://github.com/dmlc/xgboost/issues/6297.
